### PR TITLE
JDK-8324243: Compilation failures in java.desktop module with gcc 14

### PR DESCRIPTION
**Pull request by thesamesam**
This fixes building with GCC 14:
* ~Cherry-pick a fix from Harfbuzz upstream~
* Apply other `-Wcalloc-transposed-args` fixes to the JDK sources

-Wcalloc-transposed-args errors out with GCC 14 as the OpenJDK build uses
-Werror.

The calloc prototype is:
```
void *calloc(size_t nmemb, size_t size);
```

So, just swap the number of members and size arguments to match the prototype, as
we're initialising 1 struct of size `sizeof(struct ...)`. GCC then sees we're not
doing anything wrong.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8324243](https://bugs.openjdk.org/browse/JDK-8324243): Compilation failures in java.desktop module with gcc 14 (**Bug** - P4)

### Reviewers
 * [Julian Waters](https://openjdk.org/census#jwaters) (@TheShermanTanker - Committer)
 * [Magnus Ihse Bursie](https://openjdk.org/census#ihse) (@magicus - **Reviewer**)
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [ef698d82](https://git.openjdk.org/jdk/pull/17506/files/ef698d82830e3a2374e67f27069831a6df00adb6)
 * [Phil Race](https://openjdk.org/census#prr) (@prrace - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/17506/head:pull/17506` \
`$ git checkout pull/17506`

Update a local copy of the PR: \
`$ git checkout pull/17506` \
`$ git pull https://git.openjdk.org/jdk.git pull/17506/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 17506`

View PR using the GUI difftool: \
`$ git pr show -t 17506`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/17506.diff">https://git.openjdk.org/jdk/pull/17506.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/17506#issuecomment-1902689938)

### CODE REVIEW

**Comment by TheShermanTanker** [comment]
@thesamesam Unfortunately we can't fix the harfbuzz sources, since they are third party code. Your changes to harfbuzz code wouldn't be accepted by the client team :(

**Comment by thesamesam** [comment]
> @thesamesam Unfortunately we can't fix the harfbuzz sources, since they are third party code. Your changes to harfbuzz code wouldn't be accepted by the client team :(

Bleh. I don't suppose any exception can be made given it's a cherry-pick (i.e. my changes there are _not_ original, they're upstream, just not in a release)?

If not, I can just drop that bit, and we can wait until a new Harfbuzz release is made and propagates into OpenJDK.

**Comment by TheShermanTanker** [comment]
I don't believe harfbuzz has been bumped into a specific commit before, so I'm not sure about that. Perhaps you could wait till Phil sees this, for a conclusive answer

**Comment by thesamesam** [comment]
ok - if I have to drop it, so be it, not the end of the world. I guess it's also tricky because of cherry-picking vs bumping to some git commit as a whole. I'll wait a little bit then just drop if nobody chimes in.

**Comment by TheShermanTanker** [comment]
You'd need a issue on the tracker to connect to this Pull Request for anyone to see this (most reviewers don't go through these manually, they come from the mailing lists). Are you able to create one yourself, or should I create one for you?

**Comment by thesamesam** [comment]
Would you be so kind to file one for me please? I don't have access yet.

**Comment by TheShermanTanker** [comment]
No problem!

Just change your Pull Request title to 8324243 (The literal number) and watch the magic happen :)

By the way, your name strikes me as familiar. Have I seen you before on the gcc mailing lists?

**Comment by thesamesam** [comment]
> No problem!
> 

Many thanks!

> Just change your Pull Request title to 8324243 (The literal number) and watch the magic happen :)
> 
> By the way, your name strikes me as familiar. Have I seen you before on the gcc mailing lists?

You know, now you say it, I think we have interacted before actually - either on the GCC ML or bugzilla!

**Comment by TheShermanTanker** [comment]
> > No problem!
> 
> Many thanks!
> 
> > Just change your Pull Request title to 8324243 (The literal number) and watch the magic happen :)
> > By the way, your name strikes me as familiar. Have I seen you before on the gcc mailing lists?
> 
> You know, now you say it, I think we have interacted before actually - either on the GCC ML or bugzilla! I tend to pop up in a lot of places just because we have a lot of users with a lot of different odd and interesting environments.

Ah! I think you might be the same Sam James in my thread about adding the invalid-noreturn warning to gcc (Which ultimately didn't get anywhere). Glad to meet you again! :)

This Pull Request might be related to client libs code, but given its nature I'd consider it a build-dev issue too, in which case, have a thumbs up from me, since that is my domain.

Thanks, and welcome to OpenJDK if you're new here!

**Review by TheShermanTanker — APPROVED** [review]

**Comment by thesamesam** [comment]
Ah, yes! FWIW, I'm still disappointed about that as well (I don't have any sort of power, just tend to do a lot of triage). Do consider sending it again when stage 1 opens again after GCC 14 is out. Persistence is key with GCC, even if your change isn't an objectionable one. It took me a little while to get used to that..

File: src/java.desktop/share/native/libharfbuzz/hb-object.hh
Comment (prrace) [inline src/java.desktop/share/native/libharfbuzz/hb-object.hh:328]: Note 
(1) We don't cherry pick unless it is critical.
(2) I doubt you have ownership of harfbuzz code such that you can contribute any part of it to JDK in accordance with the OCA you signed.

Accordingly, the right thing to do here is disable the warning for this library and tool chain.

Comment (thesamesam) [reply src/java.desktop/share/native/libharfbuzz/hb-object.hh:328]: Sure, I wasn't exactly hiding that it was cherry-picked. I did not consider (2) to be an issue given OpenJDK wholesale imports releases, but w/e.

Anyway, happy to do that. Can you suggest what the idiomatic way of doing that in the OpenJDK codebase is for imported libs?

Comment (prrace) [reply src/java.desktop/share/native/libharfbuzz/hb-object.hh:328]: add the warning here[
](https://github.com/openjdk/jdk/blob/8b9bf758801400e4491326cd4c90fc117b9d97e1/make/modules/java.desktop/lib/Awt2dLibraries.gmk#L506)

Comment (prrace) [reply src/java.desktop/share/native/libharfbuzz/hb-object.hh:328]: PS make sure you include a comment which toolchain / version required this

**Review by magicus — APPROVED** [review]
Changes look good from a build perspective, but you need an ok from client as well for java.desktop changes.

File: make/modules/java.desktop/lib/Awt2dLibraries.gmk
Comment (kimbarrett) [inline make/modules/java.desktop/lib/Awt2dLibraries.gmk:514]: Inconsistent indentation.

**Review by kimbarrett — APPROVED** [review]
(I'm not a member of the relevant teams, but...)
The changes to calloc calls are plainly an improvement.  The new gcc
warnings are correct in the sense of potentially confused code, though
in all of these cases I think it's "harmless".

Just a minor whitespace issue in the build changes.

**Review by TheShermanTanker — APPROVED** [review]

**Review by magicus — APPROVED** [review]

**Review by prrace — APPROVED** [review]

**Comment by prrace** [comment]
@thesamesam you need to fix the PR title to match the JBS issue just as the bot is instructing you.
Then the bots should mark this "ready" to push.

**Comment by thesamesam** [comment]
Thanks, I hadn't gone back and read the original comment.

**Comment by TheShermanTanker** [comment]
Glad to have met you again in this Pull Request :)
/sponsor

**Comment by thesamesam** [comment]
Thank you!

**Pull request closed without merging**
